metadata.yaml

```
@@ -8,4 +8,10 @@ tags:
 requires:
   reverseproxy:
     interface: http
-series: ['xenial']
\ No newline at end of file
+resources:
+    dhparam:
+        type: file
+        filename: dhparam.pem
```

Is there a reason why the charm can't generate the dhparam file itself? Seems like way to low-level to expose to the operator.

---

pyOpenSSL does not have a function to generate dhparameters. Using subprocess kept silently failing during charm deployment (not in this charm, but I tested it in the sojobo-api charm)

---

Hm, can the easy-rsa charm generate dh params for you? I'm generating dh params in the openvpn charms without any issue, so let's sit together about this if easy-rsa can't fix it.

---

PS: isn't dhparams the responsibility of the lets-encrypt layer? I'd think any ssl connection needs those, no? Does the letsencrypt layer generate those and if not, should it?

---

They are not required for ssl to work, it just adds extra security (and is recommended). But they are allowed to be public (since they are transmitted during the exchange). So perhaps better is to add one as a file which is then copied during charm installation to a fixed directory?

---

Ok, I did not know that, awesome! That seems like a good solution, it will also speed up the deployment time.

Make sure to document the command used to generate that file, and maybe a link to SE or something that explains that reusing params is ok.

---

PS: If you want to do some more research on this, apparently, there are standardized secure sets of DH params available, might be nice to use one of those.

- https://news.ycombinator.com/item?id=9355649
- https://tools.ietf.org/html/rfc5114

---

Me neither, further research in DHParams revealed this :stuck_out_tongue: The DHParam sets seem to be to low in size, 4096 is recommended at this time. And I think the cyclic ones are used in ECDH (elliptic curve DH) ciphers, but some used ciphers do not use these (correct me if i'm wrong), but this should change in the future I recon

---

So I will use a param generated by openssl, and will document the command